File: Client/src/modules/models/bullet/MagicBullet.js
var MagicBullet = Bullet.extend({
    ctor:function(tower){
        this._super(tower);
    },
    updateBullet: function (){
        var dt = 1/GC.BATTLE.TICK_PER_SECOND;
        //cc.log("bulletLogicPoint: "+this.logicPoint.x + " " +this.logicPoint.y);
        var monsterRadius;
        if (!Helper.checkNull(this.targetMonster) && this.targetMonster.isAlive){
            this.targetLogicPoint = this.targetMonster.getLogicPoint();
            monsterRadius = this.targetMonster.hitRadius;
        }
        else {
            monsterRadius = 0;
        }
            if (Helper.getDistance(this.logicPoint,this.targetLogicPoint) - monsterRadius < this.speed*dt){
            this.logicPoint = this.targetLogicPoint;
            this.bulletUI.toTarget(this.logicPoint);

            this.buff =BuffFactory.create(GC.TARGETBUFF.ID,this.bulletTargetBuffType);
            this.buff.stat = this.stat;
            this.buff.battleId = this.battleId + "_Buff_";
            this.buff.startTimeRemaining();
            this.player.addBuff(this.buff);

            for(let i=0;i<this.player.monsterInMapList.length;i++){
                let monster = this.player.monsterInMapList[i];
                if (!Helper.checkNull(monster) && monster.isAlive){
                    let monsterRadius = monster.hitRadius;
                    if (Helper.getDistance(this.logicPoint, monster.getLogicPoint()) <=
                        this.radius + monsterRadius){
                        monster.hitBullet(this.damage);
                        monster.addBuff(this.buff);
                    }
                }
            }

            this.isAlive = false;
            //TODO something
        }
        else {
            //cc.log("speed + dt: "+this.speed + " "+dt);
            this.logicPoint = Helper.pointMoveToPoint(this.logicPoint,this.targetLogicPoint,this.speed,dt);
            this.bulletUI.updateMapPixel(this.logicPoint);
        }
        /*
        cc.log("ChasingBullet battleObjectId: "+this.battleId);
        cc.log("ChasingBullet curPos: " + this.logicPoint.x + " " + this.logicPoint.y);
        cc.log("ChasingBullet isAlive: "+this.isAlive);

         */

    },

})

File: Server/src/model/Bullet/MagicBullet.java
package model.Bullet;

import model.Define.ModelDefine;
import model.Monster.Monster;
import model.Tower.AttackTower;
import model.buff.Buff;
import model.buff.targetbuff.TargetBuff;

public class MagicBullet extends Bullet {
    private TargetBuff buff;
    public MagicBullet(AttackTower tower){
        super(tower, ModelDefine.BULLET_TYPE_CHASING);
    }
    public void update(double dt){
        super.update(dt);
        double monsterRadius;
        if (this.targetMonster != null && this.targetMonster.getIsAlive()){
            this.targetPos = this.targetMonster.getCurPos();
            monsterRadius = this.targetMonster.getHitRadius();
        }
        else {
            monsterRadius = 0;
        }

        if (this.curPos.getDistance(this.targetPos) - monsterRadius < this.speed*dt){
            this.curPos = this.targetPos;

            this.buff =new TargetBuff(this.targetBuffType,3);
            this.buff.setStat(this.stat);
            this.buff.setBattleObjectId(this.battleObjectId + "_Buff_");
            this.buff.startTimeRemaining();
            this.player.addBuff(this.buff);

            for(Monster monster: this.player.getMonsterInMapList()) {
                monsterRadius = monster.getHitRadius();
                if (this.curPos.getDistance(monster.getCurPos()) <= this.radius + monsterRadius) {
                    monster.hit(damage);
                    monster.addBuff(this.buff);
                }
            }
            this.isAlive = false;
        }
        else {
            this.curPos.translate(this.targetPos,this.speed,dt);
        }

//        System.out.println("ChasingBullet battleObjectId: "+this.battleObjectId);
//        System.out.println("ChasingBullet curPos: " + this.curPos.x + " " + this.curPos.y);
//        System.out.println("ChasingBullet isAlive: "+this.isAlive);

    }
}
